### DISCUSSION TITLE
How to send a correct JSON to provider_data in bot.send_invoice()?

### DISCUSSION THREAD

**Discussion by Serge-Mazur** [python-telegram-bot/python-telegram-bot#4469] [Q&A] [answered]
Here's my code:
https://pastebin.com/3x0E07ty

The issue is that the Tech Support of the Payment System says that they get `receipt=NULL` in `provider_data`. Something is wrong with the JSON.

I use `provider_data = json.dumps(payment_params)` to convert the dictionary into a JSON.

1. How to see the actual POST request that is being sent to the Payment System?
2. How to form a correct JSON in `provider_data`?

Thank you.

**Comment by Bibo-Joshi** [comment]
Hi. Your code looks fine for me at first glance. Could You run your code with log level set to debug? PTB and httpx log which data they send. That should be an indication of whether the data is sent correctly to TG. Additionally, you could try to manually make the request in the browser or via curl and see if that works better 🤔

**Reply by Serge-Mazur** [reply]
The debug level showed me this:
2024-09-10 13:23:59,719 - telegram.ext.ExtBot - DEBUG - Calling Bot API endpoint `sendInvoice` with parameters
`{'chat_id': 000000000, 'title': 'TEST Python The Language', 'description': 'Python is a high-level, general-purpose programming language.', 'payload': 'Payload-Testing', 'provider_token': '000000000000000', 'currency': 'USD', 'prices': [LabeledPrice(amount=99900, label='Material of 19910220PYTHON')], 'provider_data': '[{"receipt": {"phone": "00000000000", "items": [{"description": "Material of 19910220PYTHON", "quantity": "1.00", "amount": {"value": "999.00", "currency": "USD"}, "vat_code": 1}]}}]', 'need_name': False, 'need_phone_number': False, 'need_email': False, 'need_shipping_address': False, 'send_phone_number_to_provider': False, 'send_email_to_provider': False, 'reply_parameters': ReplyParameters(message_id=0000)}`

Could you tell me how to make my requests sendInvoice manually? Is there a tool or examples of POST requests of sendInvoice?
My searches give my nothing useful so far.

**Comment by Serge-Mazur** [comment, marked as answer]
I figured out what was going wrong with the module `requests`. I was constructing POST requests little step at a time to pin down the root cause.

This line `provider_data = json.dumps(payment_params)` returns a string with square brackets. These brackets break the POST request, so if you remove them then the API of the Payment System begins to read the POST request just fine.

This code is the fix:
`provider_data = json.dumps(payment_params)`
`provider_data = provider_data[1:-1]`

**Reply by Bibo-Joshi** [reply]
With your analysis, I also now see in line 52 a trailing comma - making the variable a tuple rather that a dict. If you remove that, you should not need your workaround at all :)

**Reply by Serge-Mazur** [reply]
Tested that and yes, you are correct. Edited my pastebin just in case, the referenced line is now 53.
Dang, that one comma gave me a lot of frustration. 😐
